File: packages/control-plane/loop_control_plane/agent_workflow.py
from __future__ import annotations

import asyncio
from datetime import UTC, datetime
from typing import Any, Literal
from uuid import UUID, uuid4

from pydantic import BaseModel, ConfigDict, Field

from loop_control_plane._app_agents import AgentRecord
from loop_control_plane.workspaces import WorkspaceError

BranchStatus = Literal["active", "staged", "merged", "abandoned"]
ChangeSetStatus = Literal[
    "draft",
    "ready_for_tests",
    "ready_for_review",
    "converted_to_release_candidate",
    "abandoned",
]
ReleaseCandidateStatus = Literal[
    "draft",
    "testing",
    "blocked",
    "ready_for_approval",
    "approved",
    "deployable",
    "superseded",
]
GateStatus = Literal["pending", "passed", "failed"]


class BranchCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str = Field(min_length=1, max_length=160)
    base_version_id: str = Field(default="production", max_length=160)


class BranchRecord(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    id: str
    agent_id: UUID
    name: str
    base_version_id: str
    created_by_user_id: str
    status: BranchStatus
    created_at: datetime
    updated_at: datetime


class ChangeSetCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    branch_id: str = Field(min_length=1, max_length=160)
    name: str = Field(min_length=1, max_length=160)
    summary: str = Field(default="", max_length=1200)
    source_type: str = Field(default="manual_edit", max_length=120)
    source_refs: list[str] = Field(default_factory=list, max_length=50)
    changed_objects: list[dict[str, Any]] = Field(default_factory=list, max_length=100)


class ChangeSetTestResult(BaseModel):
    model_config = ConfigDict(extra="forbid")

    eval_results_ref: str = Field(min_length=1, max_length=240)
    required_eval_suites: list[str] = Field(default_factory=list, max_length=25)
    passed: bool = True


class ChangeSetRecord(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    id: str
    agent_id: UUID
    branch_id: str
    name: str
    summary: str
    source_type: str
    source_refs: list[str]
    changed_objects: list[dict[str, Any]]
    status: ChangeSetStatus
    created_by_user_id: str
    created_at: datetime
    updated_at: datetime
    eval_results_ref: str | None = None
    required_eval_suites: list[str] = Field(default_factory=list)


class ReleaseCandidateCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    required_eval_suites: list[str] = Field(default_factory=list, max_length=25)
    required_approvals: list[str] = Field(
        default_factory=lambda: ["owner", "compliance"],
        max_length=25,
    )


class ReleaseCandidateGateUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    gate_id: str = Field(min_length=1, max_length=120)
    status: GateStatus
    evidence_ref: str = Field(default="", max_length=240)
    message: str = Field(default="", max_length=500)


class ReleaseCandidateApproval(BaseModel):
    model_config = ConfigDict(extra="forbid")

    approval_id: str = Field(min_length=1, max_length=120)
    comment: str = Field(default="", max_length=1200)


class ReleaseCandidateRecord(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    id: str
    agent_id: UUID
    branch_id: str
    change_set_id: str
    candidate_version_id: str
    readiness: list[dict[str, Any]]
    required_eval_suites: list[str]
    required_approvals: list[dict[str, Any]]
    status: ReleaseCandidateStatus
    created_at: datetime
    updated_at: datetime


def branch_payload(record: BranchRecord) -> dict[str, Any]:
    return record.model_dump(mode="json")


def change_set_payload(record: ChangeSetRecord) -> dict[str, Any]:
    return record.model_dump(mode="json")


def release_candidate_payload(record: ReleaseCandidateRecord) -> dict[str, Any]:
    return record.model_dump(mode="json")


def workflow_payload(
    *,
    branches: list[BranchRecord],
    change_sets: list[ChangeSetRecord],
    release_candidates: list[ReleaseCandidateRecord],
) -> dict[str, Any]:
    return {
        "branches": [branch_payload(item) for item in branches],
        "change_sets": [change_set_payload(item) for item in change_sets],
        "release_candidates": [release_candidate_payload(item) for item in release_candidates],
    }


class AgentWorkflowRegistry:
    def __init__(self) -> None:
        self._branches: dict[UUID, list[BranchRecord]] = {}
        self._change_sets: dict[UUID, list[ChangeSetRecord]] = {}
        self._release_candidates: dict[UUID, list[ReleaseCandidateRecord]] = {}
        self._lock = asyncio.Lock()

    async def list_for_agent(
        self,
        *,
        agent: AgentRecord,
    ) -> tuple[list[BranchRecord], list[ChangeSetRecord], list[ReleaseCandidateRecord]]:
        async with self._lock:
            return (
                list(self._branches.get(agent.id, [])),
                list(self._change_sets.get(agent.id, [])),
                list(self._release_candidates.get(agent.id, [])),
            )

    async def create_branch(
        self,
        *,
        agent: AgentRecord,
        body: BranchCreate,
        actor_sub: str,
    ) -> BranchRecord:
        async with self._lock:
            now = datetime.now(UTC)
            record = BranchRecord(
                id=f"br_{uuid4().hex[:12]}",
                agent_id=agent.id,
                name=body.name,
                base_version_id=body.base_version_id,
                created_by_user_id=actor_sub,
                status="active",
                created_at=now,
                updated_at=now,
            )
            self._branches.setdefault(agent.id, []).insert(0, record)
            return record

    async def create_change_set(
        self,
        *,
        agent: AgentRecord,
        body: ChangeSetCreate,
        actor_sub: str,
    ) -> ChangeSetRecord:
        async with self._lock:
            if body.branch_id not in {item.id for item in self._branches.get(agent.id, [])}:
                raise WorkspaceError(f"unknown branch: {body.branch_id}")
            now = datetime.now(UTC)
            record = ChangeSetRecord(
                id=f"cs_{uuid4().hex[:12]}",
                agent_id=agent.id,
                branch_id=body.branch_id,
                name=body.name,
                summary=body.summary,
                source_type=body.source_type,
                source_refs=body.source_refs,
                changed_objects=body.changed_objects,
                status="draft",
                created_by_user_id=actor_sub,
                created_at=now,
                updated_at=now,
            )
            self._change_sets.setdefault(agent.id, []).insert(0, record)
            return record

    async def mark_ready_for_tests(
        self,
        *,
        agent: AgentRecord,
        change_set_id: str,
    ) -> ChangeSetRecord:
        async with self._lock:
            return self._update_change_set_status(
                agent=agent,
                change_set_id=change_set_id,
                allowed={"draft"},
                status="ready_for_tests",
            )

    async def mark_ready_for_review(
        self,
        *,
        agent: AgentRecord,
        change_set_id: str,
        body: ChangeSetTestResult,
    ) -> ChangeSetRecord:
        async with self._lock:
            if not body.passed:
                raise WorkspaceError("required tests must pass before review")
            record = self._update_change_set_status(
                agent=agent,
                change_set_id=change_set_id,
                allowed={"ready_for_tests"},
                status="ready_for_review",
            )
            updated = record.model_copy(
                update={
                    "eval_results_ref": body.eval_results_ref,
                    "required_eval_suites": body.required_eval_suites,
                    "updated_at": datetime.now(UTC),
                }
            )
            self._replace_change_set(agent=agent, record=updated)
            return updated

    async def create_release_candidate(
        self,
        *,
        agent: AgentRecord,
        change_set_id: str,
        candidate_version_id: str,
        body: ReleaseCandidateCreate,
    ) -> ReleaseCandidateRecord:
        async with self._lock:
            change_set = self._find_change_set(agent=agent, change_set_id=change_set_id)
            if change_set.status != "ready_for_review":
                raise WorkspaceError("change set must be ready_for_review")
            required_eval_suites = body.required_eval_suites or change_set.required_eval_suites
            readiness = [
                {
                    "id": f"eval:{suite}",
                    "label": f"Eval suite {suite}",
                    "status": "passed",
                    "evidence_ref": change_set.eval_results_ref,
                    "message": "Required eval suite passed before release candidate creation.",
                }
                for suite in required_eval_suites
            ]
            if not readiness:
                readiness = [
                    {
                        "id": "eval:required",
                        "label": "Required eval coverage",
                        "status": "pending",
                        "evidence_ref": "",
                        "message": "Attach required eval suite evidence.",
                    }
                ]
            required_approvals = [
                {
                    "id": approval_id,
                    "state": "requested",
                    "satisfied": False,
                    "comment": "",
                    "decided_at": None,
                }
                for approval_id in body.required_approvals
            ]
            now = datetime.now(UTC)
            status: ReleaseCandidateStatus = (
                "ready_for_approval"
                if readiness and all(item["status"] == "passed" for item in readiness)
                else "testing"
            )
            record = ReleaseCandidateRecord(
                id=f"rc_{uuid4().hex[:12]}",
                agent_id=agent.id,
                branch_id=change_set.branch_id,
                change_set_id=change_set.id,
                candidate_version_id=candidate_version_id,
                readiness=readiness,
                required_eval_suites=required_eval_suites,
                required_approvals=required_approvals,
                status=status,
                created_at=now,
                updated_at=now,
            )
            self._release_candidates.setdefault(agent.id, []).insert(0, record)
            self._replace_change_set(
                agent=agent,
                record=change_set.model_copy(
                    update={
                        "status": "converted_to_release_candidate",
                        "updated_at": now,
                    }
                ),
            )
            return record

    async def update_gate(
        self,
        *,
        agent: AgentRecord,
        release_candidate_id: str,
        body: ReleaseCandidateGateUpdate,
    ) -> ReleaseCandidateRecord:
        async with self._lock:
            record = self._find_release_candidate(
                agent=agent,
                release_candidate_id=release_candidate_id,
            )
            found = False
            readiness: list[dict[str, Any]] = []
            for gate in record.readiness:
                if gate.get("id") == body.gate_id:
                    found = True
                    readiness.append(
                        {
                            **gate,
                            "status": body.status,
                            "evidence_ref": body.evidence_ref,
                            "message": body.message,
                        }
                    )
                else:
                    readiness.append(gate)
            if not found:
                readiness.append(
                    {
                        "id": body.gate_id,
                        "label": body.gate_id,
                        "status": body.status,
                        "evidence_ref": body.evidence_ref,
                        "message": body.message,
                    }
                )
            status: ReleaseCandidateStatus = "testing"
            if any(item.get("status") == "failed" for item in readiness):
                status = "blocked"
            elif readiness and all(item.get("status") == "passed" for item in readiness):
                status = "ready_for_approval"
            updated = record.model_copy(
                update={
                    "readiness": readiness,
                    "status": status,
                    "updated_at": datetime.now(UTC),
                }
            )
            self._replace_release_candidate(agent=agent, record=updated)
            return updated

    async def approve(
        self,
        *,
        agent: AgentRecord,
        release_candidate_id: str,
        body: ReleaseCandidateApproval,
        actor_sub: str,
    ) -> ReleaseCandidateRecord:
        async with self._lock:
            record = self._find_release_candidate(
                agent=agent,
                release_candidate_id=release_candidate_id,
            )
            if any(item.get("status") == "failed" for item in record.readiness):
                raise WorkspaceError("failed readiness gates block approval")
            approvals: list[dict[str, Any]] = []
            found = False
            now = datetime.now(UTC)
            for approval in record.required_approvals:
                if approval.get("id") == body.approval_id:
                    found = True
                    approvals.append(
                        {
                            **approval,
                            "state": "approved",
                            "satisfied": True,
                            "actor_sub": actor_sub,
                            "comment": body.comment,
                            "decided_at": now.isoformat(),
                        }
                    )
                else:
                    approvals.append(approval)
            if not found:
                raise WorkspaceError(f"unknown approval: {body.approval_id}")
            status: ReleaseCandidateStatus = (
                "deployable"
                if approvals and all(item.get("satisfied") for item in approvals)
                else "approved"
            )
            updated = record.model_copy(
                update={
                    "required_approvals": approvals,
                    "status": status,
                    "updated_at": now,
                }
            )
            self._replace_release_candidate(agent=agent, record=updated)
            return updated

    def _find_change_set(
        self,
        *,
        agent: AgentRecord,
        change_set_id: str,
    ) -> ChangeSetRecord:
        for record in self._change_sets.get(agent.id, []):
            if record.id == change_set_id:
                return record
        raise WorkspaceError(f"unknown change set: {change_set_id}")

    def _find_release_candidate(
        self,
        *,
        agent: AgentRecord,
        release_candidate_id: str,
    ) -> ReleaseCandidateRecord:
        for record in self._release_candidates.get(agent.id, []):
            if record.id == release_candidate_id:
                return record
        raise WorkspaceError(f"unknown release candidate: {release_candidate_id}")

    def _replace_change_set(
        self,
        *,
        agent: AgentRecord,
        record: ChangeSetRecord,
    ) -> None:
        items = self._change_sets.get(agent.id, [])
        for index, item in enumerate(items):
            if item.id == record.id:
                items[index] = record
                return
        raise WorkspaceError(f"unknown change set: {record.id}")

    def _replace_release_candidate(
        self,
        *,
        agent: AgentRecord,
        record: ReleaseCandidateRecord,
    ) -> None:
        items = self._release_candidates.get(agent.id, [])
        for index, item in enumerate(items):
            if item.id == record.id:
                items[index] = record
                return
        raise WorkspaceError(f"unknown release candidate: {record.id}")

    def _update_change_set_status(
        self,
        *,
        agent: AgentRecord,
        change_set_id: str,
        allowed: set[ChangeSetStatus],
        status: ChangeSetStatus,
    ) -> ChangeSetRecord:
        record = self._find_change_set(agent=agent, change_set_id=change_set_id)
        if record.status not in allowed:
            raise WorkspaceError(
                f"change set {change_set_id} cannot move from {record.status} to {status}"
            )
        updated = record.model_copy(update={"status": status, "updated_at": datetime.now(UTC)})
        self._replace_change_set(agent=agent, record=updated)
        return updated

File: apps/studio/src/lib/agent-workflow.ts
import { cpJson, type UxWireupClientOptions } from "@/lib/ux-wireup";

export type BranchStatus = "active" | "staged" | "merged" | "abandoned";
export type ChangeSetStatus =
  | "draft"
  | "ready_for_tests"
  | "ready_for_review"
  | "converted_to_release_candidate"
  | "abandoned";
export type ReleaseCandidateStatus =
  | "draft"
  | "testing"
  | "blocked"
  | "ready_for_approval"
  | "approved"
  | "deployable"
  | "superseded";
export type GateStatus = "pending" | "passed" | "failed";

export interface AgentBranch {
  id: string;
  agent_id: string;
  name: string;
  base_version_id: string;
  created_by_user_id: string;
  status: BranchStatus;
  created_at: string;
  updated_at: string;
}

export interface AgentChangeSet {
  id: string;
  agent_id: string;
  branch_id: string;
  name: string;
  summary: string;
  source_type: string;
  source_refs: string[];
  changed_objects: Array<Record<string, unknown>>;
  status: ChangeSetStatus;
  created_by_user_id: string;
  created_at: string;
  updated_at: string;
  eval_results_ref: string | null;
  required_eval_suites: string[];
}

export interface ReleaseCandidateGate {
  id: string;
  label: string;
  status: GateStatus;
  evidence_ref: string | null;
  message: string;
}

export interface ReleaseCandidateApproval {
  id: string;
  state: string;
  satisfied: boolean;
  comment: string;
  decided_at: string | null;
  actor_sub?: string;
}

export interface AgentReleaseCandidate {
  id: string;
  agent_id: string;
  branch_id: string;
  change_set_id: string;
  candidate_version_id: string;
  readiness: ReleaseCandidateGate[];
  required_eval_suites: string[];
  required_approvals: ReleaseCandidateApproval[];
  status: ReleaseCandidateStatus;
  created_at: string;
  updated_at: string;
}

export interface AgentWorkflow {
  branches: AgentBranch[];
  change_sets: AgentChangeSet[];
  release_candidates: AgentReleaseCandidate[];
}

type AgentWorkflowClientOptions = UxWireupClientOptions & {
  allowFixture?: boolean;
};

type ChangeSetClientOptions = AgentWorkflowClientOptions & {
  fallbackChangeSet?: AgentChangeSet;
};

type ReleaseCandidateClientOptions = AgentWorkflowClientOptions & {
  fallbackReleaseCandidate?: AgentReleaseCandidate;
};

const LOCAL_NOW = new Date(0).toISOString();

export function localAgentWorkflow(agentId: string): AgentWorkflow {
  return {
    branches: [
      {
        id: "br_local_refund",
        agent_id: agentId,
        name: "draft/refund-policy-fix",
        base_version_id: "production",
        created_by_user_id: "local-builder",
        status: "active",
        created_at: LOCAL_NOW,
        updated_at: LOCAL_NOW,
      },
    ],
    change_sets: [
      {
        id: "cs_local_refund",
        agent_id: agentId,
        branch_id: "br_local_refund",
        name: "Use current refund policy",
        summary:
          "Collected behavior and eval changes for current refund-policy citations.",
        source_type: "failed_eval",
        source_refs: ["eval/refund/current-policy"],
        changed_objects: [
          {
            type: "behavior",
            id: "behavior.refund_policy",
            summary: "Cite May 2026 policy before refund window.",
          },
        ],
        status: "ready_for_review",
        created_by_user_id: "local-builder",
        created_at: LOCAL_NOW,
        updated_at: LOCAL_NOW,
        eval_results_ref: "eval/run/refund-core/green",
        required_eval_suites: ["refund-core"],
      },
    ],
    release_candidates: [
      {
        id: "rc_local_refund",
        agent_id: agentId,
        branch_id: "br_local_refund",
        change_set_id: "cs_local_refund",
        candidate_version_id: "ver_local_refund",
        readiness: [
          {
            id: "eval:refund-core",
            label: "Eval suite refund-core",
            status: "passed",
            evidence_ref: "eval/run/refund-core/green",
            message:
              "Required eval suite passed before release candidate creation.",
          },
        ],
        required_eval_suites: ["refund-core"],
        required_approvals: [
          {
            id: "owner",
            state: "requested",
            satisfied: false,
            comment: "",
            decided_at: null,
          },
          {
            id: "compliance",
            state: "requested",
            satisfied: false,
            comment: "",
            decided_at: null,
          },
        ],
        status: "ready_for_approval",
        created_at: LOCAL_NOW,
        updated_at: LOCAL_NOW,
      },
    ],
  };
}

export async function listAgentWorkflow(
  agentId: string,
  opts: AgentWorkflowClientOptions = {},
): Promise<AgentWorkflow> {
  return cpJson<AgentWorkflow>(
    `/agents/${encodeURIComponent(agentId)}/workflow`,
    {
      ...opts,
      fallback:
        opts.allowFixture === true
          ? localAgentWorkflow(agentId)
          : { branches: [], change_sets: [], release_candidates: [] },
    },
  );
}

export async function createAgentBranch(
  agentId: string,
  input: { name: string; base_version_id?: string },
  opts: AgentWorkflowClientOptions = {},
): Promise<AgentBranch> {
  const fallback = {
    ...localAgentWorkflow(agentId).branches[0]!,
    ...input,
    id: `br_local_${Date.now()}`,
    status: "active" as const,
    updated_at: new Date().toISOString(),
  };
  return cpJson<AgentBranch>(
    `/agents/${encodeURIComponent(agentId)}/branches`,
    {
      ...opts,
      method: "POST",
      body: input,
      allowFallback: opts.allowFixture === true,
      fallback,
    },
  );
}

export async function createAgentChangeSet(
  agentId: string,
  input: {
    branch_id: string;
    name: string;
    summary?: string;
    source_type?: string;
    source_refs?: string[];
    changed_objects?: Array<Record<string, unknown>>;
  },
  opts: AgentWorkflowClientOptions = {},
): Promise<AgentChangeSet> {
  const fallback = {
    ...localAgentWorkflow(agentId).change_sets[0]!,
    ...input,
    id: `cs_local_${Date.now()}`,
    status: "draft" as const,
    updated_at: new Date().toISOString(),
    eval_results_ref: null,
    required_eval_suites: [],
  };
  return cpJson<AgentChangeSet>(
    `/agents/${encodeURIComponent(agentId)}/change-sets`,
    {
      ...opts,
      method: "POST",
      body: input,
      allowFallback: opts.allowFixture === true,
      fallback,
    },
  );
}

export async function markChangeSetReadyForTests(
  agentId: string,
  changeSetId: string,
  opts: ChangeSetClientOptions = {},
): Promise<AgentChangeSet> {
  return cpJson<AgentChangeSet>(
    `/agents/${encodeURIComponent(agentId)}/change-sets/${encodeURIComponent(
      changeSetId,
    )}/ready-for-tests`,
    {
      ...opts,
      method: "POST",
      allowFallback: opts.allowFixture === true,
      fallback: {
        ...(opts.fallbackChangeSet ??
          localAgentWorkflow(agentId).change_sets[0]!),
        status: "ready_for_tests",
        updated_at: new Date().toISOString(),
      },
    },
  );
}

export async function markChangeSetReadyForReview(
  agentId: string,
  changeSetId: string,
  input: {
    eval_results_ref: string;
    required_eval_suites?: string[];
    passed?: boolean;
  },
  opts: ChangeSetClientOptions = {},
): Promise<AgentChangeSet> {
  return cpJson<AgentChangeSet>(
    `/agents/${encodeURIComponent(agentId)}/change-sets/${encodeURIComponent(
      changeSetId,
    )}/ready-for-review`,
    {
      ...opts,
      method: "POST",
      body: input,
      allowFallback: opts.allowFixture === true,
      fallback: {
        ...(opts.fallbackChangeSet ??
          localAgentWorkflow(agentId).change_sets[0]!),
        status: "ready_for_review",
        eval_results_ref: input.eval_results_ref,
        required_eval_suites: input.required_eval_suites ?? ["refund-core"],
        updated_at: new Date().toISOString(),
      },
    },
  );
}

export async function createReleaseCandidate(
  agentId: string,
  changeSetId: string,
  input: { required_eval_suites?: string[]; required_approvals?: string[] },
  opts: AgentWorkflowClientOptions = {},
): Promise<AgentReleaseCandidate> {
  return cpJson<AgentReleaseCandidate>(
    `/agents/${encodeURIComponent(agentId)}/change-sets/${encodeURIComponent(
      changeSetId,
    )}/release-candidates`,
    {
      ...opts,
      method: "POST",
      body: input,
      allowFallback: opts.allowFixture === true,
      fallback: {
        ...localAgentWorkflow(agentId).release_candidates[0]!,
        id: `rc_local_${Date.now()}`,
        change_set_id: changeSetId,
        required_eval_suites: input.required_eval_suites ?? ["refund-core"],
        status: "ready_for_approval",
        updated_at: new Date().toISOString(),
      },
    },
  );
}

export async function approveReleaseCandidate(
  agentId: string,
  releaseCandidateId: string,
  input: { approval_id: string; comment?: string },
  opts: ReleaseCandidateClientOptions = {},
): Promise<AgentReleaseCandidate> {
  const base =
    opts.fallbackReleaseCandidate ??
    localAgentWorkflow(agentId).release_candidates[0]!;
  const approvals = base.required_approvals.map((approval) =>
    approval.id === input.approval_id
      ? {
          ...approval,
          state: "approved",
          satisfied: true,
          comment: input.comment ?? "",
          decided_at: new Date().toISOString(),
        }
      : approval,
  );
  return cpJson<AgentReleaseCandidate>(
    `/agents/${encodeURIComponent(
      agentId,
    )}/release-candidates/${encodeURIComponent(releaseCandidateId)}/approve`,
    {
      ...opts,
      method: "POST",
      body: input,
      allowFallback: opts.allowFixture === true,
      fallback: {
        ...base,
        required_approvals: approvals,
        status: approvals.every((approval) => approval.satisfied)
          ? "deployable"
          : "approved",
        updated_at: new Date().toISOString(),
      },
    },
  );
}
